### PR TITLE
Create a documentation page with all options

### LINKED ISSUE #6155: `pylint` documentation doesn't have page listing all command line options
### Question

Hello, I am searching command-line options and examples of how to use it for `pylint` in the documentation but it's very hard and not organized. I want to help document and reorganize commad-line options and maybe config file like `pyproject.toml` and `pylintrc` documentation also can get some love :)

### Documentation for future user

from my experience, `pip` docs have good command line section/topic, maybe `pylint` can copy the way command line options organize in `pip`

[pip commadline docs](https://pip.pypa.io/en/latest/cli/)

### Additional context

I love to help, any advice on where I can see all available options and descriptions in the codebase?

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to ``CONTRIBUTORS.txt`` if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :scroll: Docs          |

## Description

Closes #6155

### CODE REVIEW

**Comment by DanielNoord** [comment]
See example here: https://pylint--6346.org.readthedocs.build/en/6346/user_guide/configuration/all-options.html

File: doc/exts/pylint_options.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+def _get_all_options(linter: PyLinter) -> OptionsDataDict:
+    """Get all options registered to a linter and return the data."""
+    all_options: OptionsDataDict = defaultdict(list)
+    for checker in sorted(
+        linter.get_checkers(),
+        key=lambda x: (
+            x.name != "master",
+            getmodule(x).__name__.startswith("pylint.extensions."),  # type: ignore[union-attr]
+            x.name,
+        ),
```
Comment (Pierre-Sassoulas) [inline doc/exts/pylint_options.py:57]: I think we should modify ``__gt__`` in BaseChecker instead. (https://github.com/PyCQA/pylint/blob/main/pylint/checkers/base_checker.py#L50)

File: doc/exts/pylint_options.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+) -> str:
+    checker_string = get_rst_title(f"``{checker.capitalize()}`` Checker", "^")
+
+    toml_doc = tomlkit.document()
+    pylint_tool_table = tomlkit.table(is_super_table=True)
+    toml_doc.add(tomlkit.key(["tool", "pylint"]), pylint_tool_table)
+
+    checker_table = tomlkit.table()
+
+    for option in sorted(options, key=lambda x: x):
```
Comment (Pierre-Sassoulas) [inline doc/exts/pylint_options.py:84]: Isn't this the default ?
```suggestion
    for option in sorted(options):
```

File: doc/exts/pylint_options.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+.. raw:: html
+
+   </details>
+"""
+
+    return checker_string
+
+
+def _write_options_page(options: OptionsDataDict, linter: PyLinter) -> None:
+    """Create or overwrite the file for each message."""
```
Comment (Pierre-Sassoulas) [inline doc/exts/pylint_options.py:139]:
```suggestion
    """Create or overwrite the file with an up-to-date documentation of all the messages."""
```
I understood it as "for each message we overwrite the file with a documentation for this message" at first.

File: doc/user_guide/configuration/naming-styles.rst [hunk 4]
```diff
@@ -0,0 +1,192 @@
+.. -*- coding: utf-8 -*-
```
Comment (Pierre-Sassoulas) [inline doc/user_guide/configuration/naming-styles.rst:1]: Is this file auto-generated, or should I review it ?

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Looks pretty good already, this documentation was much needed for a long time 🎉

File: doc/exts/pylint_options.py (hunk 1 above)
Comment (DanielNoord) [reply doc/exts/pylint_options.py:57]: I'll do this in a separate PR and block this by that PR.

File: doc/exts/pylint_options.py (hunk 2 above)
Comment (DanielNoord) [reply doc/exts/pylint_options.py:84]: Oops, that should be `x.name` I think. Let me double-check.

File: doc/exts/pylint_options.py (hunk 3 above)
Comment (DanielNoord) [reply doc/exts/pylint_options.py:139]: Incorrect copy 😅 

Should not be anything about `messages`.

File: doc/user_guide/configuration/naming-styles.rst (hunk 4 above)
Comment (DanielNoord) [reply doc/user_guide/configuration/naming-styles.rst:1]: It's the old page that I moved to a new file (see the diff). I changed nothing but the title

File: doc/exts/pylint_options.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    toml_string = "\n".join(f"   {i}" for i in tomlkit.dumps(toml_doc).split("\n"))
+    checker_string += f"""
+.. raw:: html
+
+   <details>
+   <summary><a>Example configuration section</a></summary>
+
+**Note:** Only ``pylint.tool`` is required, the section title is not. These are the default values.
+
+.. code::
```
Comment (gunungpw) [inline doc/exts/pylint_options.py:126]: Maybe use `.. code:: toml` for toml syntax highlighting. If none I will use python syntax highlighting
if none:
image
if use `.. code::toml`:
image

Comment (DanielNoord) [reply doc/exts/pylint_options.py:126]: Oh I think I got a warning previously that that directive wasn't recognise. Let me recheck!

Comment (gunungpw) [reply doc/exts/pylint_options.py]: `.. code:: toml`
`.. code:: toml:` not valid I think
edit: already fix 👍

Comment (DanielNoord) [reply doc/exts/pylint_options.py]: Yes, thanks for the tip!

File: doc/exts/pylint_options.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    toml_doc = tomlkit.document()
+    pylint_tool_table = tomlkit.table(is_super_table=True)
+    toml_doc.add(tomlkit.key(["tool", "pylint"]), pylint_tool_table)
+
+    checker_table = tomlkit.table()
+
+    for option in sorted(options, key=lambda x: x.name):
+        checker_string += get_rst_title(f"--{option.name}", '"')
+        checker_string += f"\nDescription: *{option.optdict.get('help')}*\n\n"
+        checker_string += f"Default: *{option.optdict.get('default')}*\n\n\n"
```
Comment (gunungpw) [inline doc/exts/pylint_options.py:80]: if the default value is empty string `""`, the docs will just render as  `**`
image
maybe we should use `metavar` to replace this, when the default value is empty string.
I check many options that have default value `""` either accept `regex`, `msg id`, or some multiple choice
There are 13 case for this

Comment (gunungpw) [reply doc/exts/pylint_options.py:80]: also for formatting maybe we can use
[Definition List](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#definition-lists)  or [Field list](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#field-lists) to get identation for separation between `Description` `Default` and their value

Comment (gunungpw) [reply doc/exts/pylint_options.py:80]: # Definition list
```python
        checker_string += f"\nDescription:\n  *{option.optdict.get('help')}*\n\n"
        if option.optdict.get("default") == "":
            checker_string += f"Default:\n  ``{option.optdict.get('metavar')}``\n\n\n"
        else:
            checker_string += f"Default:\n  ``{option.optdict.get('default')}``\n\n\n"
```
image

Comment (DanielNoord) [reply doc/exts/pylint_options.py]: I have changed it now print `""` if that is the default value. Let me know what you think!

Comment (gunungpw) [reply doc/exts/pylint_options.py]: looks good 👍, also another question I see a lot of similarities with `Technical Reference/Pylint features` do we need to do something about it?

Comment (DanielNoord) [reply doc/exts/pylint_options.py]: We should probably revisit the Technical Reference at some point as there is indeed a bit of duplication there. However, I don't think normal users are likely to go there to look for configuration help so I think it is good to create this new section.

**Review by Pierre-Sassoulas — APPROVED** [review]
👍 Thank you for reviewing @gunungpw

**Pull request merged**